dance/pipeline.py

```diff
@@ -848,15 +853,137 @@ def flatten_dict(d, *, parent_key="", sep="_"):
+    root_path = DANCEDIR if root_path is None else root_path
+    save_directory = f"{root_path}/{save_directory}"
+    path = f"{root_path}/{path}"
+    config = OmegaConf.load(path)
+    dict_config = DictConfig(config)
+    nums = dict_config[tune_mode]
+    subsets = generate_combinations_with_required_elements(nums, required_indexes)
+    configs = []
+    command_str = "#!/bin/bash\nlog_dir=" + log_dir + "\nmkdir -p ${log_dir}\n"
+    for index, subset in enumerate(subsets):
+        config_copy = dict_config.copy()
+        config_copy[tune_mode] = subset
+        configs.append(config_copy)
+        save_path = f"{save_directory}/subset_{index}_{tune_mode}_tuning_config.yaml"
+        if save_config:
+            OmegaConf.save(config_copy, save_path)
+        count = reduce(mul, [len(p["include"]) if "include" in p else 1 for p in subset])
+        config_dir = os.path.relpath(os.path.dirname(save_path), os.path.dirname(os.path.join(root_path, file_path)))
+        command_str = command_str + f"python {file_path} --config_dir={config_dir}/subset_{index}_ --count={count} > {log_dir}/{index}.log 2>&1 &\n"
+    return command_str, configs
+
+
+def get_step3_yaml(conf_save_path="examples/tuning/cta_svm/config_yamls/params/",
+                   conf_load_path="examples/tuning/cta_svm/cell_type_annotation_default_params.yaml",
+                   result_load_path="examples/tuning/cta_svm/results/pipeline/best_test_acc.csv",
+                   required_funs=["SetConfig"], required_indexes=[-1], root_path=None):
+    """Generate the configuration file of step 3 based on the results of step 2.
+
+    Parameters
+    ----------
+    conf_save_path
+        Storage directory of the configuration file generated in step 3
+    conf_load_path
+        Parameter search range of all preprocessing functions under a specific algorithm task
+    result_load_path
+        The storage path of the result of step 2
+    required_funs
+        Required functions in step 3
+    required_indexes
+        Location of required functions in step 3
+    root_path
+        root path of all paths
+
+    """
+    DANCEDIR = Path(__file__).resolve().parent.parent
+    root_path = DANCEDIR if root_path is None else root_path
+    conf_save_path = os.path.join(root_path, conf_save_path)
+    conf_load_path = os.path.join(root_path, conf_load_path)
+    result_load_path = os.path.join(root_path, result_load_path)
+    conf = OmegaConf.load(conf_load_path)
+    result = pd.read_csv(result_load_path)
+    columns = sorted([col for col in result.columns if col.startswith("pipeline")])
+    pipeline_names = result.loc[:, columns].values
+    count = 0
+    for row in pipeline_names:
+        pipeline = []
+        row = [i for i in row]
+        for i, f in zip(required_indexes, required_funs):
+            row.insert(i, f)
+        for x in row:
+            for k in conf.pipeline:
+                if k["target"] == x:
+                    pipeline.append(k)
```

Comment:
The corresponding pipeline was found from the default parameter search range file